delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/files/batch/BatchFiles.scala

```diff
@@ -0,0 +1,81 @@
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.model.FileCommand._
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.model.FileRejection.CopyRejection
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.model._
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.routes.CopyFileSource
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.{FetchFileStorage, FileResource}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.model.StorageType
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.StorageFileRejection.CopyFileRejection
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import ch.epfl.bluebrain.nexus.delta.sdk.identities.model.Caller
+import ch.epfl.bluebrain.nexus.delta.sdk.implicits._
+import ch.epfl.bluebrain.nexus.delta.sdk.projects.FetchContext
+import ch.epfl.bluebrain.nexus.delta.sdk.projects.model.ProjectContext
+import ch.epfl.bluebrain.nexus.delta.sourcing.model.ResourceRef
+
+trait BatchFiles {
+  def copyFiles(
+      source: CopyFileSource,
+      dest: CopyFileDestination
+  )(implicit c: Caller): IO[NonEmptyList[FileResource]]
+}
+
+object BatchFiles {
+  def mk(
+      fetchFileStorage: FetchFileStorage,
+      fetchContext: FetchContext[FileRejection],
+      evalFileCommand: CreateFile => IO[FileResource],
+      batchCopy: BatchCopy
+  )(implicit uuidF: UUIDF): BatchFiles = new BatchFiles {
+
+    private val logger = Logger[BatchFiles]
+
+    implicit private val kamonComponent: KamonMetricComponent = KamonMetricComponent(entityType.value)
+
+    override def copyFiles(source: CopyFileSource, dest: CopyFileDestination)(implicit
+        c: Caller
+    ): IO[NonEmptyList[FileResource]] = {
+      for {
+        pc                            <- fetchContext.onCreate(dest.project)
+        (destStorageRef, destStorage) <- fetchFileStorage.fetchAndValidateActiveStorage(dest.storage, dest.project, pc)
+        destFilesAttributes           <- batchCopy.copyFiles(source, destStorage).adaptError { case e: CopyFileRejection =>
+                                           CopyRejection(source.project, dest.project, destStorage.id, e)
+                                         }
+        fileResources                 <- createFileResources(pc, dest, destStorageRef, destStorage.tpe, destFilesAttributes)
+      } yield fileResources
+    }.span("copyFiles")
+
+    private def createFileResources(
+        pc: ProjectContext,
+        dest: CopyFileDestination,
+        destStorageRef: ResourceRef.Revision,
+        destStorageTpe: StorageType,
+        destFilesAttributes: NonEmptyList[FileAttributes]
+    )(implicit c: Caller): IO[NonEmptyList[FileResource]] =
+      destFilesAttributes.traverse { destFileAttributes =>
+        for {
+          iri      <- generateId(pc)
+          command   =
+            CreateFile(iri, dest.project, destStorageRef, destStorageTpe, destFileAttributes, c.subject, dest.tag)
```

Comment:
Not sure if we want to preserve the original tag.
It would be nice in some way to know that this file has been copied from another